### ISSUE TITLE
Platypus awareness campaign

### ISSUE THREAD

**Issue by plicease** [plicease/FFI-Platypus#28]
Things to do to get the word out about FFI / Platypus:
- <strike>Remove `FFI` from the top of metacpan.org, or failing that SEE ALSO it / reimplement in something more modern (ticket: https://rt.cpan.org/Ticket/Display.html?id=101880)</strike> (<b>update</b> now maintained CPAN module with SEE ALSOs pointing to `FFI::Platypus`, `FFI::Raw` and `Win32::API`)
- Get FFI::Platypus on the libffi home page (ticket: https://github.com/atgreen/libffi/issues/131)
- Convince other CPAN authors to port their existing modules to Platypus (one success here: https://github.com/gugod/Phash-FFI/pull/1), another good candidate would be the zmq interface (ported)
- blog about FFI::Platypus (examples: http://blogs.perl.org/users/graham_ollis/2015/01/practical-ffi-with-platypus.html , http://npointsolutions.blogspot.in/2015/02/forgin-function-interface-with-perl.html)
- Add additional language plugins.  When I created the Rust plugin I had two rust programmers star it almost immediately.  This is double plus good since it also gets the word out about Perl 5.
- Talk about Platypus at local and regional events.  I will be giving a presentation at DCBPW (http://dcbpw.org/dcbpw2015/talk/6008) and  YAPC::NA (http://www.yapcna.org/yn2015/talk/6072).
- Favorite FFI::Platypus in metacpan.org (with your PAUSE id)
- Recruit people to take over some of the language plugins / custom type modules, to get people invested in FFI infrastructure
- get the ppm build for this thing working. Right now it gets bogged down trying to install PkgConfig.pm see #37 Work arounds in PkgConfig.pm would be acceptable.

**Comment by plicease** [comment]
@awwaiid @merrilymeredith @lungching ideas / suggestions / help on the awareness campaign would be welcome :)

**Comment by awwaiid** [comment]
Should we recommend a naming scheme, or an anti-naming scheme? So like I want to fork Audio::PortAudio. Should I:
* Get maintenance over it and switch it to use FFI::Platypus
* Create a new lib called Audio::PortAudio::FFI
* Create a new lib called Audio::PortAudio::Platypus

**Comment by merrilymeredith** [comment]
You probably saw that I just updated my repo with a big note pointing over here, in case anyone still finds their way to that.

I can blog about it this weekend, maybe a more general write-up, then tweet about it and see about getting it in something like Perl Weekly.  I'll try to find something to make a new FFI module for too.

**Comment by plicease** [comment]
@awwaiid Entirely my opinion, but: I've been using the `::FFI` to signal "this is a module implemented in some form of FFI for Perl (but should eventually be Platyous if not already :P ) and may have advantages over equivalent versions that either lack a suffix or have an `::XS` suffix ".  Sort of like `::XS` means "this module is implemented in XS so might be faster than the Pure perl version that lacks the a suffix."  I think that if you are implementing a module and it happens to use FFI you do not need a `::FFI` suffix.  In fact I think if we can infect CPAN with FFI stuff that just works without people noticing it that would be better than waving the flag *every* time.

There is sort of a precedent for this as well, examples: `ZMQ::FFI` and `Crypt::ZCert`.

All that said, I think if you can get maint on `Audio::PortAudio` and port it to Platypus I wouldn't rename it.  If you can't get maint, I don't think there is any problem in implementing a compatible `Audio::PortAudio::FFI`.  Not sure how I feel about `Audio::PortAudio::Platypus` :P though I am a big fan of the animal.

@merrilymeredith I did see that, thanks!  If you don't mind mentioning me `@plicease` (or maybe we should use a hashtag #perlffi or something) in any tweets so I see it and post a link to any blogs here.  Sometimes I peruse https://github.com/ffi/ffi/wiki/projects-using-ffi to look for easy pickings.  That is how I came up with `FFI::TinyCC`.  Maybe that link is old news.

**Comment by plicease** [comment]
Another couple of bullet points I had inside my head but didn't mention when I opened this issue:

  - Favorite `FFI::Platypus` in metacpan.org
  - Recruit people to take over some of the language plugins / custom type modules, to get people invested in FFI infrastructure 

The rationale behind language and custom type modules to give people a way of building on the FFI ecosystem without having to get patches accepted to core Platypus.  The author of `FFI::Raw` was usually pretty good about accepting patches, but it was a bottle neck and a single point of failure.  Things that are really important could eventually make it into the core.  I've documented at least a couple that are available for adoption.

These modules are ripe:

  - `FFI::Platypus::Lang::Rust`
  - `FFI::Platypus::Lang::CPP`
  - `FFI::Platypus::Lang::CPP::Demangle::XS`
  - `FFI::Platypus::Type::StringArray`

These come with Platypus but there is not reason they couldn't be spun off:

  - `FFI::Platypus::Type::StringPointer`
  - `FFI::Platypus::Type::PointerSizeBuffer`

This one I am keeping for sentimental reasons.

  - `FFI::Platypus::Lang::Pascal`

**Comment by plicease** [comment]
Not to shabby:

http://perltricks.com/article/149/2015/2/6/What-s-new-on-CPAN---January-2015

**Comment by plicease** [comment]
I noticed that the pretricks entry referred to using FFI to call Python... which i dunno might be possible through C or something, but not really practical at the moment at least.  I think part of the confusion might have been my preamble:

```
Platypus provides a method for creating interfaces to machine code libraries. This implementation uses libffi, a library that provides Foreign Function Interfaces for a number of other languages, including Ruby and Python. Platypus can be used in stand alone scripts, or to create Perl extensions for CPAN. There are a number of reasons why you might want to write extensions with FFI instead of XS:
```
I can see how this could be confusing so it should be re-written to be more clear.

**Comment by awwaiid** [comment]
Same article also referred to FFI::Me.

**Comment by plicease** [comment]
yeah I made an attempt to get the author of FFI::Me on board with Platypus, but haven't heard anything beyond an initial exchange (take a look at the gituhb repo for Lchmod if you want to see the PR)... I find that FFI::Me doesn't really add much to FFI::Raw except for an alternate syntax, but to each their own.

btw- @awwaiid you asked me about structs on Tuesday and I hadn't done any work on it, but I started in the master branch there is now a FFI::Platypus::Record that does rubyish layouts.  It doesn't appear to be as hard as I imagined it to be.  I will be doing some portability testing that I hope won't change my mind.  #21 if you want to see my notes.  Basically I need to implement strings for it (not entirely trivial) and I should be shipping it to CPAN shortly.

**Comment by merrilymeredith** [comment]
I started to make a plan and ... I think I'm looking at writing a *pair* of posts now, one getting a simple library on Alien and another building a basic FFI::Platypus interface to it.  I have to educate myself on both kits, so it'd be a blog-along for that.

**Comment by plicease** [comment]
- get the ppm build for this thing working.  Right now it gets bogged down trying to install PkgConfig.pm see #37 Work arounds in PkgConfig.pm would be acceptable.

**Comment by plicease** [comment]
I am now the owner of the `FFI` and `FFI::Library` module.  I am going to keep it an interface to `ffcall` so that we can compare and contrast with `libffi` based solutions.  I've fixed a bug on 64 bit systems (the reason for much of the rest on the testers matrix), and added SEE ALSOs to Platypus and `FFI::Raw`.

**Comment by calid** [comment]
FYI I've started migrating `ZMQ::FFI` to `FFI::Platypus`.  xsub attach is a killer feature and resulted in some pretty dramatic speedups for me locally.

**Comment by plicease** [comment]
@calid awesome.  let me know if you run into any road blocks that I can help with.

**Comment by calid** [comment]
I finished the initial port.  I still have a lot of cleanup and optimizing left to do, but even in its 'rough' form the results are already pretty wild.  I used the zeromq weather station example to compare performance:

```
# FFI::Raw Implementation

$ time perl wuserver.pl & 
$ time perl wuclient.pl
Collecting updates from weather station...
Average temperature for zipcode '10001 ' was 21F
 
real 1m22.818s  <=====
user 0m0.070s
sys 0m0.023s
 
 
# FFI::Platypus Implementation

$ time perl wuserver.pl & 
$ time perl wuclient.pl
Collecting updates from weather station...
Average temperature for zipcode '10001 ' was 32F
 
real 0m22.645s  <=====
user 0m0.053s
sys 0m0.023s 
```

So just switching to FFI::Platypus shaved a minute off the example.  

I profiled the server sending out a million messages and (predictably) `zmq_send` was the primary bottleneck in both cases.  But the difference between the two implementations is pretty stark:

```
# FFI::Raw Implementation

$self->_zmq3_ffi->{zmq_send}->($self->_socket, $msg, $length, $flags)
# spent 19.9s making 1000000 calls to FFI::Raw::__ANON__[FFI/Raw.pm:94], avg 20µs/call
# spent 5.72s making 2000000 calls to FFI::Raw::coderef, avg 3µs/call
# spent 2.90s making 1000000 calls to ZMQ::FFI::ZMQ3::Socket::_zmq3_ffi, avg 3µs/call
 
# FFI::Platypus Implementation
 
zmq3_send($self->_socket, $msg, $length, $flags)
# spent 1.76s making 1000000 calls to ZMQ::FFI::ZMQ3::Socket::zmq3_send, avg 2µs/call

sub ZMQ::FFI::ZMQ3::Socket::zmq3_send; # xsub 
```

Once I actually optimize the Perl code I should be able to get performance on par with the Python bindings, and definitely faster than the XS bindings.

Nice work!

Gist with the weather station example code and results: https://gist.github.com/17df5bcfb81c83786d6f.git

**Comment by plicease** [comment]
@calid thanks!  and this looks great.  If you don't mind I will probably use this as an example when I talk about Platypus at DC Baltimore PW (http://dcbpw.org/dcbpw2015/talk/6008) and possibly YAPC::NA.  Let me know if you can get the final version faster than the XS version because that would be a wonderful counter example to the "XS is faster" critics.

**Comment by calid** [comment]
Sure, absolutely.  I've updated the results using the optimized/cleaned up code and formalized the gist a bit.  I also lightly optimized the example code (e.g. don't declare vars in loop) and made it more 1:1 between FFI and XS so they're on a level playing field.  The XS example code is posted now as well.

Now for the results.  First, the bad news: Achieving parity with the Python bindings was a little too ambitious... Perl subroutine overhead ultimately kills performance compared to the Python bindings.

But here's the good news: While the XS came out _slightly_ ahead of the FFI, they're so close that once you factor in the vagaries of timing processes it's effectively equivalent performance.  Which means we get to have our cake and eat it too:  the flexibility of FFI with the performance of XS (heh, maybe that should be the FFI::Platypus tagline :) ).

To put a finer point on it, here's the profiling results:

##### ZMQ::LibZMQ3 / XS

```
zmq_send($socket, $string, -1); 
# spent 1.23s making 1000000 calls to ZMQ::LibZMQ3::zmq_send, avg 1µs/call
```

Statements | Time on Line | Calls | Time in Subs |
---------------|------------------|--------|----------------- | 
1000000 | 6.26s | 1000000 | 1.23s	| 

##### ZMQ::FFI / FFI::Platypus

```
zmq_send($socket, $msg, $length, $flags)
# spent 1.33s making 1000000 calls to ZMQ::FFI::ZMQ3::Socket::zmq_send, avg 1µs/call
```

Statements | Time on Line | Calls | Time in Subs |
---------------|------------------|--------|----------------- | 
1000000 | 6.37s | 1000000 | 1.33s	| 

So as you can see once the FFI function is loaded as an xsub the performance for the native stuff is basically the same, and then it just comes down to how well you optimize the Perl code itself.  And both the XS and FFI suffer from the same xsub overhead (Time on Line), which I don't think there's a solution for (outside of hacking on Perl itself :) )

**Comment by calid** [comment]
Regarding getting the word out, I'll be giving a talk about `FFI::Platypus` and my experience migrating `ZMQ::FFI` at Chicago.PM on April 23

**Comment by calid** [comment]
@plicease more bad news. The assertion that FFI xsubs are just fast (or close enough not to matter) isn't holding up to further scrutiny. I Benchmarked just plain FFI vs XS sends and XS came out 50% faster:

```perl
my $r = timethese 10_000_000, {
    'XS'  => sub {
        die 'xs send error ' if -1 == zmq_send($xs_socket, 'ohhai', 5, 0);
    },

    'FFI' => sub {
        die 'ffi send error' if -1 == zmqffi_send($ffi_socket, 'ohhai', 5, 0);
    },
};
cmpthese($r);
```

```
Benchmark: timing 10000000 iterations of FFI, XS...
       FFI:  4 wallclock secs ( 3.31 usr +  0.01 sys =  3.32 CPU) @ 3012048.19/s (n=10000000)
        XS:  2 wallclock secs ( 2.16 usr +  0.00 sys =  2.16 CPU) @ 4629629.63/s (n=10000000)

         Rate   FFI    XS     C
FFI 3012048/s    --  -35%  -82%
XS  4629630/s   54%    --  -73%
C  16835017/s  559%  364%    --
```

I benchmarked the equivalent C manually and added it to the results just so it's easy to compare a baseline.

I'm a bit confused what the slowdown could be since the profiler indicated the two xsubs have almost identical performance.

**Comment by calid** [comment]
Ok, some good news.  Remember how I said being as fast as the Python bindings was "too ambitions"?  Wellll, I looked at the Python example code and found this:

```python
# Process 5 updates
total_temp = 0
for update_nbr in range(5)
    ...
```

Python is only doing 5 updates where C and Perl are doing 100!  When you time the example using 100 updates instead it comes out to `1m40s` for the Python vs `13s` for ZMQ::FFI/FFI::Platypus.  So yeah, we blow the doors off the Python bindings :)

After further testing I'm also reasonably comfortable I have a handle on how FFI::Platypus compares to XS (about 30% slower, but still not enough to justify ditching the convenience of ffi).  I've added a Performance blurb to the ZMQ::FFI pod which you can read at calid/zmq-ffi@8f9bb98f, so let me know if you agree with how I've phrased the performance trade offs.

I'll probably release the FFI::Platypus version of ZMQ::FFI tonight or tomorrow assuming no issues come up.

Cheers

**Comment by plicease** [comment]
@calid 30% sounds about right with what I am seeing in my benchmarks.  it's good to see about the same thing in a practical real life example.  We may be able to improve further on that by optimizing Platypus, but I agree the benefits of using Platypus and FFI are already there.  I appreciate the work that you have done (and the shout out in your POD) it will give me something to talk about.

**Comment by plicease** [comment]
@calid ... and yes totally agree with what you've said and how you said it.

**Comment by plicease** [comment]
I'll be giving a talk at YAPC::NA on FFI:

http://www.yapcna.org/yn2015/talk/6072

If you are a fan of FFI and you are going to be in Salt Lake for YAPC::NA (or DCBPW) be sure to find me.

**Comment by plicease** [comment]
https://twitter.com/plicease/status/591655725820153856

I would love to have a better Rust example for the YAPC::NA edition of my talk ... maybe even as a lightning talk.  I think there is a lot of potential here, but my rust-foo is pretty basic.

**Comment by plicease** [comment]
@felliott asked if there was a place on IRC for FFI / Platypus and because I am not active on IRC as far as I know the answer is not yet.  However, I think it would be good to have.

**Comment by felliott** [comment]
Does Alien::Base have an IRC channel? That may be a good place for it
similar to how #web-simple is the home of the Moo ecosystem.

**Comment by plicease** [comment]
discussing irc here: (and on #alien)
https://github.com/Perl5-Alien/Alien-Base/pull/126
tldr: considering `#native` as a combined Alien / FFI / Platypus discussion

**Comment by plicease** [comment]
This here have been kind of slow lately, but I intend to make some progress on both `FFI::Platypus` and `Alien::Base` this year.  Blog entry here:

http://blogs.perl.org/users/yuki_kimoto/2016/05/ffiplatypus-is-interesting-it-seems-like-real-ffi-module-for-perl-5.html

Drew some attention on github and metacpan in the form of stars and +1s and some discussion on IRC.

I think the primary thing that Platypus needs is just some more bindings, and especially by people to drive development of the API.

Also worth mentioning is that @mohawk2 wrote a Perl 6 inspired `NativeCall` interface:

https://metacpan.org/release/NativeCall

Which I like a lot better than `FFI::Platypus::Declare`.  I intend to try it with some simple libraries.

**Comment by plicease** [comment]
There is a Platypus package in Debian unstable.

https://packages.debian.org/sid/libffi-platypus-perl

This has also revealed an issue with bigendian 64bit systems (like s390x) #67 which I am close to resolving.
